prql-compiler/src/sql/srq/ast.rs

```diff
@@ -0,0 +1,170 @@
+        with: Rel,
+        filter: Expr,
+    },
+
+    Distinct,
+    Except {
+        bottom: Rel,
+        distinct: bool,
+    },
+    Intersect {
+        bottom: Rel,
+        distinct: bool,
+    },
+    Union {
+        bottom: Rel,
+        distinct: bool,
+    },
+}
+
+impl<Rel> SqlTransform<Rel> {
+    pub fn as_str(&self) -> &str {
+        match self {
+            SqlTransform::Super(t) => t.as_ref(),
+            _ => self.as_ref(),
+        }
+    }
+
+    pub fn into_super_and<T, F: FnOnce(rq::Transform) -> Result<T, rq::Transform>>(
+        self,
+        f: F,
+    ) -> Result<T, Self> {
+        self.into_super()
+            .and_then(|t| f(t).map_err(SqlTransform::Super))
+    }
+}
+
+pub trait SrqFold<RelIn, RelOut, SuperIn, SuperOut>: RqFold {
+    fn fold_rel(&mut self, rel: RelIn) -> Result<RelOut>;
+
+    fn fold_super(&mut self, sup: SuperIn) -> Result<SuperOut>;
+
+    fn fold_sql_transforms(
+        &mut self,
+        transforms: Vec<SqlTransform<RelIn, SuperIn>>,
+    ) -> Result<Vec<SqlTransform<RelOut, SuperOut>>> {
+        transforms
+            .into_iter()
+            .map(|t| self.fold_sql_transform(t))
+            .try_collect()
+    }
+
+    fn fold_sql_transform(
+        &mut self,
+        transform: SqlTransform<RelIn, SuperIn>,
+    ) -> Result<SqlTransform<RelOut, SuperOut>> {
+        fold_sql_transform::<RelIn, RelOut, SuperIn, SuperOut, _>(self, transform)
+    }
+}
+
+pub fn fold_sql_transform<
+    RelIn,
+    RelOut,
+    SuperIn,
+    SuperOut,
+    F: ?Sized + SrqFold<RelIn, RelOut, SuperIn, SuperOut>,
+>(
+    fold: &mut F,
+    transform: SqlTransform<RelIn, SuperIn>,
+) -> Result<SqlTransform<RelOut, SuperOut>> {
+    Ok(match transform {
+        SqlTransform::Super(t) => SqlTransform::Super(fold.fold_super(t)?),
+
+        SqlTransform::From(rel) => SqlTransform::From(fold.fold_rel(rel)?),
+        SqlTransform::Join { side, with, filter } => SqlTransform::Join {
+            side,
+            with: fold.fold_rel(with)?,
+            filter: fold.fold_expr(filter)?,
+        },
+
+        SqlTransform::Distinct => SqlTransform::Distinct,
+        SqlTransform::Union { bottom, distinct } => SqlTransform::Union {
+            bottom: fold.fold_rel(bottom)?,
+            distinct,
+        },
+        SqlTransform::Except { bottom, distinct } => SqlTransform::Except {
+            bottom: fold.fold_rel(bottom)?,
+            distinct,
+        },
+        SqlTransform::Intersect { bottom, distinct } => SqlTransform::Intersect {
+            bottom: fold.fold_rel(bottom)?,
+            distinct,
+        },
+        SqlTransform::Select(v) => SqlTransform::Select(v),
+        SqlTransform::Filter(v) => SqlTransform::Filter(v),
+        SqlTransform::Aggregate { partition, compute } => {
+            SqlTransform::Aggregate { partition, compute }
+        }
+        SqlTransform::Sort(v) => SqlTransform::Sort(v),
+        SqlTransform::Take(v) => SqlTransform::Take(v),
```

Comment:
Many / most of these could just be

```suggestion
        _ => transform
```

...I think, but no great issue, and maybe explicit forces us to consider each case